### ISSUE TITLE
Bookmark Branch Porter: add "tags", "keywords" and "lastModified" properties to JSON export

### ISSUE THREAD

**Issue by ericmarceau-rogers** [igorlogius/igorlogius#34]
The JSON file created by "Export Branch (JSON)" is missing some bookmark properties namely "**tags**", "**keywords**", and "**lastModified**" which are contained in the browser built-in bookmark exporter. 

#### Missing **tags** and **keywords** properties

|Fx built-in| BBP addon|
|---|---|
|![SNAPSOT__Firefox_TagsAndKeyword](https://github.com/user-attachments/assets/0870231e-9ebe-483b-895b-d61773c3510e) | ![SNAPSOT__BMPorter_TagsAndKeyword](https://github.com/user-attachments/assets/daf224ec-6d73-4c62-8ce2-110e962e50b7) |

#### Missing **lastModified** property

|Fx built-in| BBP addon|
|---|---|
| ![SNAPSHOT__Firefox__LastModified](https://github.com/user-attachments/assets/1abf0d72-948b-4efd-bb55-a763ffd68e4d) | ![SNAPSHOT__BMPorter__LastModified](https://github.com/user-attachments/assets/fbeee91d-7f3b-4cd6-b54a-8c0f127593d4) |

**Comment by igorlogius** [comment]
Thank your for the very detailed report. 

Sadly the addon can only serializes the existing [BookmarkTreeNode](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/bookmarks/BookmarkTreeNode) object provided by the addon API.

The "tags", "keyword" and "lastModified" attribute are as you can see missing here. - Likely because those are not compatible with other browsers and the API was designed to be generic as possible AFAIK. 

So, to fix this issue, mozilla will have extend the addon API to include the missing properties. A good places to ask for this are either on Bugzilla or on Connect. 

Not much else i can do here. Sorry. 

Would be nice if you could edit/clarify this in your AMO review so it doenst read like a feature that is missing. 

I'll also add a note in the description page ... so users wont be suprised that these properties are not included in the export/import.

**Comment by ericmarceau-rogers** [comment]
I understand your desire to work directly from the API.  But ...

Given that the Bookmarks "contained" in the browser are a superset of what is being covered by the current API, have you considered an alternate path to a more complete solution that deals with the "ACTUAL" Bookmarks tree?

What about using as your starting point what you currently do, allowing the end-user to select the "folder" to be exported, obtain from that selection the necessary ID of the top level item, but then extract the FULL Bookmarks file using Firefox's export function and locate, and extract, the desired subset within that full JSON to create the end-user's desired subset.

Would that be doable?

Not only would you be able to get everything, but also everything using the labels used by Firefox itself in the FULL tree export (why would those be different from what the API is making visible).

Food for thought ... and maybe a different approach?

Thank you for hearing me out!

**Comment by ericmarceau-rogers** [comment]
Bug report submitted to Mozilla:

https://bugzilla.mozilla.org/show_bug.cgi?id=1909560

**Comment by igorlogius** [comment]
>  the "ACTUAL" Bookmarks tree?
>  extract the FULL Bookmarks file using Firefox's export function and locate

Webextenions are very very restricted and have no acccess to either the "ACTUAL" bookmark tree and can also not access/use/invoke firefox built-in export/import functions.

> Thank you for hearing me out!

No problem. I'd really like to extend the addon to include these properties.

The only definitely working but not userfriendly and likely not very popular solution would be by using the native-messaging API, which allows an addon to communicate with an external helper program (python/nodejs script).  The helper program could be used to query the missing data from inside the  `places.sqlite` database and send that information back to the addon.

The issue here is that its that native messaging requires some user setup... and is a little much for most users to work with. 

If you are really interested i could look into this, but from my past experiences i remember that the `places.sqlite` is a bit of a mess ... you have to make some pretty whild queries to get certain information out of it ... not saying its not doable, just ... kind of annoying to deal with.

**Comment by igorlogius** [comment]
Since you are referencing this issue in the bugzilla ticket, i'll leave it open for now.

**Comment by igorlogius** [comment]
related: 

- https://bugzilla.mozilla.org/show_bug.cgi?id=1225916

**Comment by igorlogius** [comment]
> Thank you for the report.

> Tags are already covered by [bug 1225916](https://bugzilla.mozilla.org/show_bug.cgi?id=1225916). Unfortunately making those accessible on the API is blocked on some other needed work to rework how they are handled internally.

> For Keywords, we are considering removing them from being stored alongside bookmarks, and adding them as search engines instead. Though the exact details are still to be decided. In any case, we wouldn't want to expose these on the API at this time because of potentially removing them later. The meta bug for the work is [bug 1650874](https://bugzilla.mozilla.org/show_bug.cgi?id=1650874).

> Last modified is a possibility of something that we could expose. We need to check that Firefox Sync doesn't rely on it being set correctly (we think it doesn't but just in case), and it would need to have some boundaries (i.e. greater than dateAdded and not in the future). Note that Google Chrome does not expose the last modified date either, so this would be a slight variance on their API.

> Therefore I'm changing this bug to be explicitly about the last modified case, since tags are already covered and keywords are probably a won't fix.

Well ... guess getting one property is better than nothing :-)

**Comment by igorlogius** [comment]
Since this issue should automatically be resolved as soon as mozilla decides to add these properties, meanging there isnt anyhting todo here for me i am closing this issue for now.
